syntax = "proto3";

message DnaProto {
  float learning_rate = 1;
  repeated  EdgeProto  edges = 2;
  repeated  ConvexProto  convex =  3;

}

message EdgeProto{
    int32 from_vertex = 1;
    int32 to_vertex = 2;
    int32  depth_precedence  = 3;
    int32  scale_precedence  = 4;

  enum  EdgeType {
    CONV = 0;
    IDENTITY = 1;
  }

  message Conv {
     int32 depth_factor = 1;
     int32 filter_half_width = 2;
     int32 filter_half_height = 3;
     int32 stride_scale = 4;
  }

   EdgeType type = 5;
}

message ConvexProto{
   int32  inputs_mutable  = 1;
   int32  outputs_mutable = 2;
   int32  properties_mutable  = 3;
  repeated  int32  edges_in = 4;
  repeated  int32  edges_out = 5;
  enum  ConvexType {
    LINEAR = 0;
    BN_RELU = 1;
  }
  
   ConvexType type = 6;
}
